Unsafe derives and attributes (#3715)
rust-lang/rfcs: text/3715-unsafe-derives-and-attrs.md

# Guide-level explanation
[guide-level-explanation]: #guide-level-explanation

## Derives

When declaring a proc macro `derive`, you can add the `unsafe` parameter to the
`proc_macro_derive` attribute to indicate that the derive requires `unsafe`:

```rust
#[proc_macro_derive(DangerousTrait, unsafe)]
pub fn derive_helper_attr(_item: TokenStream) -> TokenStream {
    TokenStream::new()
}
```

Invoking this derive requires writing `#[derive(unsafe(DangerousTrait))]`.
Invoking an unsafe derive without the unsafe derive syntax will produce a
compiler error. Using the unsafe derive syntax without an unsafe derive will
trigger an "unused unsafe" lint.

A `proc_macro_derive` attribute can include both `attributes` for helper
attributes and `unsafe` to declare the derive unsafe, in any order.

If writing code that enforces `SAFETY` comments for every use of `unsafe`, you
can write the `SAFETY` comment either prior to the derive (for a single unsafe
derive) or prior to the specific `unsafe(DangerousTrait)` in a list of derives:

```rust
/// SAFETY: ...
#[derive(unsafe(DangerousTrait))]
struct SomeStruct { ... }

#[derive(

>     /// SAFETY: ...

Reviewer: This strikes me as a bit odd, is this supposed to be a line comment (that is non-doc comment)? Documentation comments only have very specific places where they are allowed. Also they represent `#[doc = "content"]` so this would be equivalent to writing
```rust
#[derive(
    #[doc = "SAFETY: ..."]
    unsafe(DangerousTrait)
...
```

- Feature Name: `unsafe_derives_and_attrs`
- Start Date: 2024-10-22
- RFC PR: [rust-lang/rfcs#3715](https://github.com/rust-lang/rfcs/pull/3715)
- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)

# Summary
[summary]: #summary

Allow declaring proc macro attributes and derive macros as unsafe, and
requiring `unsafe` to invoke them.

# Motivation
[motivation]: #motivation

Some traits place requirements on implementations that the Rust compiler cannot
verify. Those traits can mark themselves as unsafe, requiring `unsafe impl`
syntax to implement. However, trait `derive` macros cannot currently require
`unsafe`. This RFC defines a syntax for declaring and using unsafe `derive`
macros.

This RFC also defines a syntax for declaring proc macro attributes as unsafe.

# Guide-level explanation
[guide-level-explanation]: #guide-level-explanation

## Derives

> When declaring a proc macro `derive`, you can add the `unsafe` parameter to the
> `proc_macro_derive` attribute to indicate that the derive requires `unsafe`:
> 
> ```rust
> #[proc_macro_derive(DangerousTrait, unsafe)]
> pub fn derive_helper_attr(_item: TokenStream) -> TokenStream {
>     TokenStream::new()
> }
> ```

Reviewer 2: Given that we're putting the `unsafe` around the derive macro name at the use site, what do we think about matching that at the def site?  E.g.:

```rust
#[proc_macro_derive(unsafe(DangerousDeriveMacro)]
pub fn derive_helper_attr(_item: TokenStream) -> TokenStream {
    TokenStream::new()
}
```

```suggestion
When declaring a proc macro `derive`, you can wrap the derive macro name with
`unsafe(..)` in the `proc_macro_derive` attribute to indicate that the derive
requires `unsafe`:

```rust
#[proc_macro_derive(unsafe(DangerousDeriveMacro)]
pub fn derive_helper_attr(_item: TokenStream) -> TokenStream {
    TokenStream::new()
}
```
```

Reviewer 3: On one hand I like the symmetry, on the other hand (because we overloaded unsafe to have two meanings) I think this could be read as the derive macro itself discharging a safety obligation, whereas the first form makes it slightly clearer that this is just an unsafe derive.

Author: @carbotaniuman That's my feeling as well: `proc_macro_derive(unsafe(DangerousDeriveMacro))` reads to me as though the creation of it is unsafe, rather than it being a safe operation that creates an unsafe derive.

Author: I've added an alternative for this, along with rationale for why we shouldn't do it.

> Invoking this derive requires writing `#[derive(unsafe(DangerousTrait))]`.
> Invoking an unsafe derive without the unsafe derive syntax will produce a
> compiler error. Using the unsafe derive syntax without an unsafe derive will
> trigger an "unused unsafe" lint.
> 
> A `proc_macro_derive` attribute can include both `attributes` for helper
> attributes and `unsafe` to declare the derive unsafe, in any order.
> 
> If writing code that enforces `SAFETY` comments for every use of `unsafe`, you
> can write the `SAFETY` comment either prior to the derive (for a single unsafe
> derive) or prior to the specific `unsafe(DangerousTrait)` in a list of derives:
> 
> ```rust
> // SAFETY: ...
> #[derive(unsafe(DangerousTrait))]
> struct SomeStruct { ... }
> 
> #[derive(
>     // SAFETY: ...
>     unsafe(DangerousTrait),
>     // SAFETY: ...
>     unsafe(AnotherDangerousTrait),
> )]
> struct AnotherStruct { ... }
> ```

Reviewer 2: Given the distinction we're drawing to support the choice of syntax, it's probably better to use something like `DangerousDeriveMacro` rather than `DangerousTrait` in all of the examples.

This also removes the sentence about ordering with the `unsafe` parameter to match the suggestion above that we write `unsafe(DeriveMacro)` at the def-site.

```suggestion
Invoking this derive requires writing `#[derive(unsafe(DangerousDeriveMacro))]`.
Invoking an unsafe derive without the unsafe derive syntax will produce a
compiler error. Using the unsafe derive syntax without an unsafe derive will
trigger an "unused unsafe" lint.

If writing code that enforces `SAFETY` comments for every use of `unsafe`, you
can write the `SAFETY` comment either prior to the derive (for a single unsafe
derive) or prior to the specific `unsafe(DangerousDeriveMacro)` in a list of
derives:

```rust
// SAFETY: ...
#[derive(unsafe(DangerousDeriveMacro))]
struct SomeStruct { ... }

#[derive(
    // SAFETY: ...
    unsafe(DangerousDeriveMacro),
    // SAFETY: ...
    unsafe(AnotherDangerousDeriveMacro),
)]
struct AnotherStruct { ... }
```
```

Reviewer 2: (There remain other uses below that we'd probably want to update as well to match.)

## Attributes

When declaring a proc macro attribute, you can add the `unsafe` parameter to
the `proc_macro_attribute` attribute to indicate that the attribute requires
`unsafe`:

```rust
#[proc_macro_attribute(unsafe)]
pub fn dangerous(_attr: TokenStream, item: TokenStream) -> TokenStream {
    item
}
```

Invoking an unsafe attribute requires the unsafe attribute syntax:
`#[unsafe(dangerous)]`.

> If writing code that enforces `SAFETY` comments for every use of `unsafe`, you
> can write the `SAFETY` comment immediately prior to the attribute:

Reviewer 2: ```suggestion
You can write the `SAFETY` comment immediately prior to the attribute:
```

"What, you mean there's code somewhere that doesn't require that?"  I'd prefer to not hint that this expectation is somehow not the default.  Probably we can just leave this predicate out entirely.

Author: @traviscross It is, in a very practical sense, *not* currently the default: there's no warn-by-default lint for it. I'm all for changing that, but meanwhile, it'd be inaccurate to imply that it is currently required everywhere.

Reviewer 2: Josh and I talked, but for others here, by wanting to remove the predicate, I don't mean to suggest by this language that it is the default or is required.

If we say, "you can write X here," I don't think that implies that you must write "X" at all.  I read it instead as, "if you're going to write an X, which is entirely your option, you can write it here."

What I don't want to suggest is that writing such safety comments is unusual, and the fussy language referencing extrinsic requirements for doing this makes it feel that way to me.

I proposed wording along the lines of "when adding a safety comment, you can write it here" that both Josh and I liked, and that I think he's planning to update the text to reflect.

```rust
// SAFETY: ...
#[unsafe(dangerous)]
```

# Rationale and alternatives
[rationale-and-alternatives]: #rationale-and-alternatives

This RFC proposes the synax `#[derive(unsafe(DangerousTrait))]`. We could,
instead, put the `unsafe` on the outside: `#[unsafe(derive(DangerousTrait))]`.

Some rationale for putting it on the inside:
- This encourages minimizing the scope of the `unsafe`, isolating it to a

>   single trait.

Reviewer 4: one could make `#[unsafe(derive(Foo, Bar))]` invalid syntax.

Author: I don't think we should use a syntax that naturally leads people to assume this would work, and then complain when they do the natural thing that seems like it should work.

Nonetheless, I've captured this in the alternatives section.

> - This encourages minimizing the scope of the `unsafe`, isolating it to a
>   single trait.

Reviewer 2: ```suggestion
- This encourages minimizing the scope of the `unsafe`, isolating it to a
  single derive macro.
```

- This allows writing all traits to derive within a single `#[derive(...)]`, if
  desired. Putting the `unsafe` on the outside requires separate `derive`s for
  safe and unsafe derives, and potentially multiple if the derives care about

>   ordering.

Reviewer 4: afaik `#[derive(X)]` cannot see another `#[derive(Y)]` and all a derive-macro does is append an item, which is normally order-agnostic.

> [[items.decl-order]](https://doc.rust-lang.org/reference/items.html#r-items.decl-order)
>
> Items may be defined in **any order**, with the exception of `macro_rules` which has its own scoping behavior.

Author: @kennytm While I think it's good practice to make derives order-independent, it's very possible to make them order-dependent, in multiple ways. First, as you noted, the are exceptions like macros. Second proc macros can have side effects outside the scope of the generated code.

Nonetheless, I've called out that this is unusual.

> - This allows writing all traits to derive within a single `#[derive(...)]`, if
>   desired. Putting the `unsafe` on the outside requires separate `derive`s for
>   safe and unsafe derives, and potentially multiple if the derives care about
>   ordering.

Reviewer 2: ```suggestion
- This allows writing all derive macros to invoke within a single
  `#[derive(...)]`, if desired. Putting the `unsafe` on the outside requires
  separate `derive`s for safe and unsafe derives, and potentially multiple if
  the derives care about ordering.
```

> - This makes it easy to attach `SAFETY` comments to each individual trait.

Reviewer 4: Also I don't see how I also don't see how this is an advantage of `#[derive(unsafe(Foo))]` vs `#[unsafe(derive(Foo))]`. The latter is even 2 lines shorter.

```rust
#[derive(
    // SAFETY: ...
    unsafe(DangerousTrait),
    // SAFETY: ...
    unsafe(AnotherDangerousTrait),
)]
struct AnotherStruct { ... }

vs.

// SAFETY: ...
#[unsafe(derive(DangerousTrait))]
// SAFETY: ...
#[unsafe(derive(AnotherDangerousTrait))]
struct AnotherStruct { ... }
```

Author: @kennytm This goes along with the previous point about not having to split the `derive`.

> - This makes it easy to attach `SAFETY` comments to each individual trait.

Reviewer 2: ```suggestion
- This makes it easy to attach `SAFETY` comments to each individual derive
  macro.
```

> - One way to think of `derive(Trait)` is `derive` is the mechanism to invoke
>   derive macros, and `Trait` is the actual derive macro. In this sense, when
>   writing `derive(unsafe(UnsafeTrait))`, the `unsafe` is attached to the
>   specific derive macro rather than the `derive` directive.

Reviewer 2: ```suggestion
- One way to think of `derive(Macro)` is that `derive(..)` enters a context in
  which one or more derive macros can be invoked, and naming `Macro` is how we
  actually invoke the derive macro. When invoking unsafe derive macros, we have
  to wrap those with `unsafe(..)` as in `derive(unsafe(DangerousDeriveMacro))`.
```

`unsafe`:

```rust
#[proc_macro_attribute(unsafe)]
pub fn dangerous(_attr: TokenStream, item: TokenStream) -> TokenStream {
    item
}
```

Invoking an unsafe attribute requires the unsafe attribute syntax:
`#[unsafe(dangerous)]`.

If writing code that enforces `SAFETY` comments for every use of `unsafe`, you
can write the `SAFETY` comment immediately prior to the attribute:

```rust
// SAFETY: ...
#[unsafe(dangerous)]
```

# Rationale and alternatives
[rationale-and-alternatives]: #rationale-and-alternatives

This RFC proposes the synax `#[derive(unsafe(DangerousDeriveMacro))]`. We
could, instead, put the `unsafe` on the outside:
`#[unsafe(derive(DangerousDeriveMacro))]`.

Some rationale for putting it on the inside:
- This encourages minimizing the scope of the `unsafe`, isolating it to a
  single derive macro.
- This allows writing all derive macros to invoke within a single
  `#[derive(...)]`, if desired. Putting the `unsafe` on the outside requires
  separate `derive`s for safe and unsafe derives, and potentially multiple in
  the (unusual) case where the derives care about ordering.
- This makes it easy to attach `SAFETY` comments to each individual derive

>   macro.

Author: @kennytm That's true, though we could fix that in rustfmt in the future.

>   macro.

Author: I've added a note about this.

without the unsafe derive syntax will produce a compiler error. Using the
unsafe derive syntax without an unsafe derive macro will trigger an "unused
unsafe" lint.

A `proc_macro_derive` attribute can include both `attributes` for helper
attributes and `unsafe` to declare the derive unsafe, in any order.

When writing a `SAFETY` comment for each `unsafe`, you can place the `SAFETY`
comment either prior to the derive (for a single unsafe derive) or prior to the
specific `unsafe(DangerousDeriveMacro)` in a list of derives:

```rust
// SAFETY: ...
#[derive(unsafe(DangerousDeriveMacro))]
struct SomeStruct { ... }

#[derive(
    // SAFETY: ...
    unsafe(DangerousDeriveMacro),
    // SAFETY: ...
    unsafe(AnotherDangerousDeriveMacro),
)]
struct AnotherStruct { ... }
```

(Note that current rustfmt will place every derive on a line of its own if any
have a comment. That could be changed in a future style edition, but this RFC
is not making or advocating any style proposals.)

### Helper attributes

A `derive` macro can have helper attributes. You can use the following syntax
to declare a helper attribute as `unsafe`:

```rust

> #[proc_macro_derive(MyDeriveMacro(, attributes(unsafe(dangerous_helper_attr))]

Reviewer 2: ```suggestion
#[proc_macro_derive(MyDeriveMacro, attributes(unsafe(dangerous_helper_attr)))]
```

# Motivation
[motivation]: #motivation

Some traits place requirements on implementations that the Rust compiler cannot
verify. Those traits can mark themselves as unsafe, requiring `unsafe impl`
syntax to implement. However, trait `derive` macros cannot currently require
`unsafe`. This RFC defines a syntax for declaring and using unsafe `derive`
macros.

This provides value for any derive of an `unsafe trait`, ranging from standard
library unsafe traits such as `Send` and `Sync` to more complex unsafe traits
in the ecosystem. With this mechanism available, an `unsafe trait` has the
option of providing two different kinds of `derive` macros: a safe `derive`
macro that implements the `unsafe` trait in a fashion that's always safe (or
that fails if some obligation is not met), or an `unsafe` `derive` macro that
puts the safety obligation on the invoker of the `derive`.

Some examples of potential unsafe derives for `unsafe trait`s:

- `TrustedLen` and `DerefPure` in the standard library. (Currently unstable,
  but such a derive would be useful when they're stable, serving the function
  of an `unsafe impl`.)
- `pyo3::marker::Ungil` in `pyo3`, in place of the current handling of a

>   blanket impl for any `Send` type.

Reviewer 4: If these are the only motivation examples, I don't get why these *marker* traits `TrustedLen`, `TrustedStep`, `DerefPure`, `Ungil` *require* using the `#[derive]` mechanism over manual `unsafe impl`. 

For the first three the only advantages are not having to explicitly write the `where` bounds, but they have the `Iterator` or `Deref` supertraits which cannot be derived so you gotta repeat those bounds anyway.

Author: Literally any `unsafe trait` that one might wish to `derive` is a potential example. These were a few samples. If you have other examples of `unsafe trait` that you'd prefer to have cited instead of or in addition to these, I'd be happy to add them.

Reviewer 4: > Literally any `unsafe trait` that one might wish to `derive` is a potential example.

Normally if we wish to `derive` something it is because we know that it is possible to fill out the `impl` programmatically with the type's definition alone, thus saving effort to produce the boilerplate impl. 

If all the derive doing is just implementing a marker trait without anything extra there's certainly no point *from the supply-side* anyone wish to write a proc-macro for it. At least it should be derivable together with its base trait like `#[derive(Clone, Copy)]` i.e. `#[derive(Iterator, unsafe(TrustedLen))]`, but you can't & *can't* derive an Iterator.

Another reason we wish to `derive` a marker trait is to insert some compile-time checks, such as `zerocopy`'s macros. But `zerocopy`'s derive-macro checks at the same proved the conditions required by the `unsafe trait` in the first place, making those traits *safe to derive* even if *unsafe to manually implement*. 

For `libstd` the public, documented, non-sealed unsafe traits are:

1. Traits with no obvious derivable impl:
    * GlobalAlloc
    * Allocator
    * Searcher
    * ReverseSearcher
2. Marker traits
    * TrustedStep
    * TrustedLen
    * DerefPure
    * PinCoerceUnsized
3. TransmuteFrom
4. CloneToUninit

Perhaps `CloneToUninit` is the one most compatible with `#[derive]`:

```rust
#[derive(CloneToUninit)]
struct Packet<Tail: ?Sized> {
    header: Header,
    tail: Tail,
}
```

but the actual impl can satisfy `impl CloneToUninit`'s safety condition without user's intervention, meaning `#[derive(CloneToUninit)]` itself is actually safe.

```rust
unsafe impl<Tail: CloneToUninit + ?Sized> CloneToUninit for Packet<Tail> {
    unsafe fn clone_to_uninit(&self, dst: *mut u8) {
        self.header.clone_to_uninit(dst.add(offset_of_val!(self, header)));
        self.tail.clone_to_uninit(dst.add(offset_of_val!(self, tail)));
    }
}
```

So no I don't have any other examples to support this RFC, i.e. I'm still on the side of the concern "needs fleshed out use case".

- RFC PR: [rust-lang/rfcs#3715](https://github.com/rust-lang/rfcs/pull/3715)
- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)

# Summary
[summary]: #summary

Allow declaring proc macro attributes and derive macros as unsafe, and
requiring `unsafe` to invoke them.

# Motivation
[motivation]: #motivation

Some traits place requirements on implementations that the Rust compiler cannot
verify. Those traits can mark themselves as unsafe, requiring `unsafe impl`
syntax to implement. However, trait `derive` macros cannot currently require
`unsafe`. This RFC defines a syntax for declaring and using unsafe `derive`
macros.

This RFC also defines a syntax for declaring proc macro attributes as unsafe.

# Guide-level explanation
[guide-level-explanation]: #guide-level-explanation

## Derives

When declaring a proc macro `derive`, you can add the `unsafe` parameter to the
`proc_macro_derive` attribute to indicate that the derive requires `unsafe`:

```rust
#[proc_macro_derive(DangerousTrait, unsafe)]
pub fn derive_helper_attr(_item: TokenStream) -> TokenStream {
    TokenStream::new()
}
```

Invoking this derive requires writing either

> `#[unsafe(derive(DangerousTrait))]` or `#[derive(unsafe(DangerousTrait))]`.

Reviewer 5: Wondering if you've considered unsafe without parenthesis.`#[derive(unsafe DangerousTrait)]` has less visual noise, at least to my eyes.

Reviewer 6: the `unsafe` uses parenthesis syntax was already decided on in [RFC 3325](https://github.com/rust-lang/rfcs/blob/master/text/3325-unsafe-attributes.md)

Reviewer 4: `unsafe DangerousTrait` may be mistaken as a #3710-unsafe-flavored variant of the `DangerousTrait` trait, so no it shouldn't be considered.

Author: Added to the alternatives section with that rationale.

syntax to implement. However, trait `derive` macros cannot currently require
`unsafe`. This RFC defines a syntax for declaring and using unsafe `derive`
macros.

This RFC also defines a syntax for declaring proc macro attributes as unsafe.

# Guide-level explanation
[guide-level-explanation]: #guide-level-explanation

## Derives

When declaring a proc macro `derive`, you can add the `unsafe` parameter to the
`proc_macro_derive` attribute to indicate that the derive requires `unsafe`:

```rust
#[proc_macro_derive(DangerousTrait, unsafe)]
pub fn derive_helper_attr(_item: TokenStream) -> TokenStream {
    TokenStream::new()
}
```

Invoking this derive requires writing either
`#[unsafe(derive(DangerousTrait))]` or `#[derive(unsafe(DangerousTrait))]`.
(The latter syntax allows isolating the `unsafe` to a single derive within a
list of derives.) Invoking an unsafe derive without the unsafe derive syntax
will produce a compiler error. Using the unsafe derive syntax without an unsafe
derive will trigger an "unused unsafe" lint.

A `proc_macro_derive` attribute can include both `attributes` for helper
attributes and `unsafe` to declare the derive unsafe, in any order.

## Attributes

> When declaring a proc macro attribute, you can add the `unsafe` parameter to

Reviewer 7: @GnomedDev said
> Any reason the syntax for defining an unsafe derive/attribute isn't `unsafe fn`?
```rust
#[proc_macro_attribute]
pub unsafe fn dangerous(_attr: TokenStream, item: TokenStream) -> TokenStream {
    item
}
```

Reviewer 7: because that makes no sense, the unsafety is at a meta-level and there are no safety requirements on the macro function itself

Reviewer 8: Is there not? I thought the point of an unsafe derive/attribute was that it was unsafe to trigger and had safety requirements?

Reviewer 7: you're confusing the memory safety of the rustc process including the proc macro code with the safety of the generated code. this RFC is about the latter, making the macro an unsafe fn is about the former

Reviewer 8: I don't personally think it's that confusing, especially for users who don't know how proc macros are loaded and run in the current implementation of the compiler.
